spring-cloud-gcp-pubsub/src/main/java/org/springframework/cloud/gcp/pubsub/core/subscriber/PubSubSubscriberTemplate.java

```diff
@@ -104,6 +147,24 @@ public Subscriber subscribe(String subscription, MessageReceiver messageReceiver
 				returnImmediately));
 	}
 
+	@Override
+	public <T> List<Message<T>> pullAndConvert(String subscription, Integer maxMessages, Boolean returnImmediately,
+			Class<T> payloadType) {
+		List<AcknowledgeablePubsubMessage> pubsubMessages = this.pull(subscription, maxMessages, returnImmediately);
+
+		return pubsubMessages.stream().map(
+				m -> {
+					Map<String, Object> headers = this.headerMapper.toHeaders(m.getMessage().getAttributesMap());
+					headers.put(GcpPubSubHeaders.PUBLISH_TIME, m.getMessage().getPublishTime());
+					headers.put(GcpPubSubHeaders.MESSAGE_ID, m.getMessage().getMessageId());
+					headers.put(GcpPubSubHeaders.ACKNOWLEDGEMENT, m);
+					return new GenericMessage<>(
```

Comment:
You have to use here a `MessageBuilder`.
More over it would be better to use a `MessageBuilderFactory` from Spring Integration.
And one more concern: there are `id` and `timestamp` headers which end-user might would like do not have generated on the message.

In other words: you pull too much Spring Integration into this `PubSubSubscriberTemplate`.

Also your idea about `MessageHandler` as callback is also a bit awkward - this also come from Spring Integration world.

I'm not sure why you so don't like Spring Integration and don't want to postpone such a hard work there using appropriate constructions, but with this your current solution it fully looks as you are mixing all the Spring Integration concerns in one place.

Reply:
I like Spring Integration. :) I'm just a fan of the template as well, and trying to make sure that it is complete.

`MessageHandler` comes from `spring-messaging` which I thought is not strictly for SI and could be used in the template.